astartes/samplers/extrapolation/scaffold.py

```diff
@@ -1,5 +1,109 @@
+"""
+This sampler partitions the data based on the Bemis-Murcko scaffold function as implemented in RDKit.
+Bemis, G. W.; Murcko, M. A. The Properties of Known Drugs. 1. Molecular Frameworks. J. Med. Chem. 1996, 39, 2887−2893.
+Landrum, G. et al. RDKit: Open-Source Cheminformatics; 2006; https://www.rdkit.org.
+
+The goal is to cluster molecules that share the same scaffold.
+Later, these clusters will be assigned to training, validation, and testing split
+to create data splits that will measure extrapolation by testing on scaffolds
+that are not in the training set.
+
+"""
+import warnings
+from collections import defaultdict
+
+import numpy as np
+import rdkit
+from rdkit import Chem
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
 from astartes.samplers import AbstractSampler
+from astartes.utils.warnings import NoMatchingScaffold
 
 
 class Scaffold(AbstractSampler):
-    pass
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not isinstance(self.X[0], str) and not isinstance(
+            self.X[0], rdkit.Chem.rdchem.Mol
+        ):
+            msg = "Scaffold class requires input X to be an iterable of SMILES strings"
+            raise TypeError(msg)
+
+    def _sample(self):
+        """Implements the Scaffold sampler to identify clusters via a molecule's Bemis-Murcko scaffold."""
+        scaffold_to_indices = self.scaffold_to_smiles(self.X)
+
+        cluster_indices = np.zeros(len(self.X), dtype=np.int8)
+        # give each cluster an arbitrary ID
+        for cluster_id, (scaffold, indices) in enumerate(scaffold_to_indices.items()):
+            if scaffold == "":
+                warnings.warn(
+                    f"No matching scaffold was found for the {len(indices)} "
+                    f"molecules corresponding to indices {indices}",
+                    NoMatchingScaffold,
+                )
+            for idx in indices:
+                cluster_indices[idx] = cluster_id
+
+        self._samples_clusters = cluster_indices
+
+    def scaffold_to_smiles(self, mols):
+        """
+        Computes scaffold for each smiles string and returns a mapping from scaffolds to sets of smiles.
+
+        Params:
+            mols: A list of smiles strings or RDKit molecules.
+
+        Returns:
+            A dictionary mapping each unique scaffold to all smiles (or smiles indices) which have that scaffold.
+        """
+        scaffolds = defaultdict(set)
+        for i, mol in enumerate(mols):
+            scaffold = self.generate_bemis_murcko_scaffold(mol)
+            scaffolds[scaffold].add(i)
+
+        return scaffolds
+
+    def str_to_mol(self, string, explicit_hydrogens=False):
```

Comment:
Similar comment to below -- explicit_hydrogens will always be false. We either need to check it, or remove this.